### PR TITLE
WT-7356 Implement bulk load for tiered tables

### PR DESCRIPTION
**Pull request by keitharnoldsmith**
Add support for bulk load for tiered cursors.  

Passes the `test_bulk*.py` tests with the tiered cursor hooks.  (Or it did before WT-7173.  WT-7440 will fix that.)

### CODE REVIEW

File: src/tiered/tiered_cursor.c [hunk 1]
```diff
@@ -1118,7 +1180,7 @@ __wt_curtiered_open(WT_SESSION_IMPL *session, const char *uri, WT_CURSOR *owner,
 
     /* Check whether the exclusive open for a bulk load succeeded. */
     if (bulk && ret == EBUSY)
-        WT_ERR_MSG(session, EINVAL, "bulk-load is only supported on newly created trees");
+        ret = EINVAL;
```
Comment (keitharnoldsmith) [inline src/tiered/tiered_cursor.c:1183]: This error message was copied from the LSM code. But the python tests expect no message on this failure.

**Review by ddanderson — APPROVED** [review]
LGTM.

File: src/tiered/tiered_cursor.c [hunk 2]
```diff
@@ -1062,6 +1062,69 @@ __curtiered_next_random(WT_CURSOR *cursor)
...
+    /* Bulk cursors only support insert and close. */
+    __wt_cursor_set_notsup(cursor);
+    cursor->insert = __curtiered_insert_bulk;
+    cursor->close = __wt_curtiered_close;
+
+    WT_ASSERT(session, curtiered->cursors == NULL);
+    WT_ERR(__wt_calloc_def(session, WT_TIERED_MAX_TIERS, &curtiered->cursors));
+
+    /* Open a bulk cursor on the local tier. */
+    dhandle = tiered->tiers[WT_TIERED_INDEX_LOCAL].tier;
```
Comment (sueloverso) [inline src/tiered/tiered_cursor.c:1114]: Probably worth asserting dhandle != NULL here.

File: src/tiered/tiered_cursor.c (hunk 1 above)
Comment (sueloverso) [reply src/tiered/tiered_cursor.c:1184]: This is kind of sad. We definitely would prefer to give the user more information than less. Is there any way to change python's expectations instead? @ddanderson any thoughts?

**Review by sueloverso — APPROVED** [review]
LGTM with a couple minor comments/questions.

File: src/tiered/tiered_cursor.c (hunk 1 above)
Comment (keitharnoldsmith) [reply src/tiered/tiered_cursor.c:1184]: We do return an error message (which the python test expects) in the similar case where an application tries to open a bulk cursor on a non-empty table. So we could certainly do the same in this case.

Of course, then we should be returning that message from the other cursor types as well, which would be the real work.    I'll create a new ticket for that.

File: src/tiered/tiered_cursor.c (hunk 2 above)
Comment (keitharnoldsmith) [reply src/tiered/tiered_cursor.c:1114]: Done

File: src/tiered/tiered_cursor.c (hunk 1 above)
Comment (ddanderson) [reply src/tiered/tiered_cursor.c:1185]: You can write tests to "expect" to see an error message in Python code.  Something like:
```
               errmsg = '/.*bulk-load is only supported.*/'
                self.assertRaisesWithMessage(wiredtiger.WiredTigerError,
                    lambda: self.session.open_cursor(uri, config), errmsg)
```

**Pull request merged**